### PR TITLE
Add handling of hash in dynamic route

### LINKED ISSUE #7858: Hashes appear to break dynamic routing
image

image

### PR DESCRIPTION
**Pull request by ijjk**
This fixes `onlyAHashChange` from preventing the dynamic route from being refreshed on mount for auto exported dynamic routes

Fixes https://github.com/zeit/next.js/issues/7858

### CODE REVIEW

File: packages/next-server/lib/router/router.ts [hunk 1]
```diff
@@ -56,6 +56,7 @@ export default class Router implements BaseRouter {
   clc: ComponentLoadCancel
   pageLoader: any
   _bps: BeforePopStateCallback | undefined
+  _prevUrl?: string
```
Comment (Timer) [inline packages/next-server/lib/router/router.ts]: This should be marked private right?
```suggestion
  private _prevUrl?: string
```

Comment (ijjk) [reply packages/next-server/lib/router/router.ts]: This is removed now, reverted to working fix

File: packages/next-server/lib/router/router.ts [hunk 2]
```diff
@@ -280,11 +285,8 @@ export default class Router implements BaseRouter {
         method = 'replaceState'
       }
 
-      // @ts-ignore pathname is always a string
-      const route = toRoute(pathname)
-      const { shallow = false } = options
-
       if (isDynamicRoute(route)) {
+        this.parsedDynamicRoute = true
```
Comment (timneutkens) [inline packages/next-server/lib/router/router.ts]: This seems like a strange state 🤔 It'll not be reset after the first time it happens for subsequent route changes it'll always be true

Comment (ijjk) [reply packages/next-server/lib/router/router.ts]: Changed how it's being handled to move the dynamic values parsing above the hash change check so this is no longer used.

**Comment by timneutkens** [comment]
Closing this as talked about with @timer, we'll do a different fix.

**Pull request closed without merging**